File: working_hours/calculator.py
from collections import defaultdict
from datetime import date, timedelta

from .models import ClockEvent, DayRecord, Session


def compute(events: list[ClockEvent]) -> dict[tuple[str, str], dict]:
    """Group events by employee, pair consecutive timestamps into sessions.

    Returns {(emp_id, name): {'dept': str, 'days': {date: DayRecord}}}.
    Consecutive events per employee per day are paired as clock-in / clock-out.
    An odd number of events on a day produces an incomplete DayRecord.
    """
    by_emp: dict[tuple, dict[date, list]] = defaultdict(lambda: defaultdict(list))
    emp_dept: dict[tuple, str] = {}

    for e in events:
        key = (e.emp_id, e.name)
        emp_dept[key] = e.dept
        by_emp[key][e.timestamp.date()].append(e.timestamp)

    result: dict[tuple, dict] = {}
    for key, date_map in by_emp.items():
        days: dict[date, DayRecord] = {}
        for d, timestamps in date_map.items():
            timestamps.sort()
            sessions = [
                Session(timestamps[i], timestamps[i + 1])
                for i in range(0, len(timestamps) - 1, 2)
            ]
            incomplete = len(timestamps) % 2 == 1
            total = sum((s.duration for s in sessions), timedelta())
            days[d] = DayRecord(
                sessions=sessions,
                total=total,
                incomplete=incomplete,
                dangling=timestamps[-1] if incomplete else None,
            )
        result[key] = {"dept": emp_dept[key], "days": days}
    return result

File: packages/server/src/domain/calculator.ts
import type { ClockEvent, DayRecord, EmployeeData, Session } from './models.js'

function dateKey(d: Date): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

export function compute(events: ClockEvent[]): EmployeeData {
  const byEmp = new Map<string, { dept: string; days: Map<string, Date[]> }>()

  for (const e of events) {
    const key = `${e.empId}|${e.name}`
    if (!byEmp.has(key)) byEmp.set(key, { dept: e.dept, days: new Map() })
    const emp = byEmp.get(key)!
    emp.dept = e.dept
    const dk = dateKey(e.timestamp)
    if (!emp.days.has(dk)) emp.days.set(dk, [])
    emp.days.get(dk)!.push(e.timestamp)
  }

  const result: EmployeeData = {}
  for (const [key, { dept, days }] of byEmp) {
    const daysOut: Record<string, DayRecord> = {}
    for (const [dk, timestamps] of days) {
      timestamps.sort((a, b) => a.getTime() - b.getTime())
      const sessions: Session[] = []
      for (let i = 0; i < timestamps.length - 1; i += 2) {
        sessions.push({ clockIn: timestamps[i], clockOut: timestamps[i + 1] })
      }
      const incomplete = timestamps.length % 2 === 1
      const total = sessions.reduce((acc, s) => acc + (s.clockOut.getTime() - s.clockIn.getTime()), 0)
      daysOut[dk] = {
        sessions,
        total,
        incomplete,
        dangling: incomplete ? timestamps[timestamps.length - 1] : null,
      }
    }
    result[key] = { dept, days: daysOut }
  }
  return result
}
